### PR TITLE
[FSDP2] Auto-exclude non-floating frozen Params4bit from fully_shard to prevent QLoRA crash

### PR DESCRIPTION
**Pull request by roycho96**
# What does this PR do?

Auto-excludes incompatible BnB `Params4bit` from FSDP2 sharding via `ignored_params` to prevent QLoRA crash.

Fixes [#3983](https://github.com/huggingface/accelerate/issues/3983)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/accelerate/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/accelerate/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/accelerate/tree/main/docs#writing-documentation---specification).
- [x] Did you write any new necessary tests?

## Who can review?

@SunMarc

### CODE REVIEW

**Comment by roycho96** [comment]
> Do we get the same results as with `bnb_4bit_quant_storage` being a float ? Did you test on a e2e example to compare both ? Thanks a lot !

@SunMarc Thanks for the review!

I ran an e2e comparison on single GPU (Qwen3-0.6B + QLoRA + SFT, same seed/hyperparams):

| Step | uint8 (this PR) | bf16 quant_storage (official) |
|------|----------------|-------------------------------|
| 1 | 1.2914 | 1.2914 |
| 2 | 1.2571 | 1.2516 |
| 3 | 1.2310 | 1.2281 |
| 4 | 1.1131 | 1.1202 |
| 5 | 1.0541 | 1.0563 |
| 6 | 0.9757 | 0.9793 |
| 7 | 0.9656 | 0.9687 |
| 8 | 0.9125 | 0.9066 |
| **avg** | **1.1001** | **1.1003** |

Loss curves are nearly identical (< 0.6% difference at any step). 
The minor variations are expected from floating-point ordering 
differences in the two quant_storage representations. 
The dequantized values are equivalent.

**Comment by roycho96** [comment]
> Another question I have is that if we put those into ignored param, will fsdp stil manage to shard correctly the model across multiple gpus since most of params are Params4bit ?

You're right. With most parameters being Params4bit and excluded  from sharding, the base weights are replicated on each GPU. FSDP  still shards the trainable LoRA parameters and gradients, but the memory benefit from weight sharding is limited in this path.
The primary goal of this PR is crash prevention. Currently, users with default QLoRA config (`bnb_4bit_quant_storage` unset) get a crash on FSDP2. I think this is a common QLoRA setup.
For users who need full base weight sharding (e.g. 70B+ models), the path with `bnb_4bit_quant_storage=bf16` remains the right choice and is unaffected by this PR.

File: src/accelerate/utils/fsdp_utils.py [hunk 1]
```diff
@@ -657,14 +657,31 @@ def fsdp2_prepare_model(accelerator, model: torch.nn.Module) -> torch.nn.Module:
...
+            if (not p.requires_grad) and (not p.is_floating_point()) and (not p.is_complex())
+        }
+        if incompatible_params4bit:
+            ignored = set(fsdp2_kwargs.get("ignored_params", set()))
+            fsdp2_kwargs["ignored_params"] = ignored | incompatible_params4bit
+            if accelerator.is_main_process:
+                logger.info(
+                    f"Found {len(incompatible_params4bit)} non-floating frozen Params4bit. "
+                    "Excluding from FSDP2 sharding to prevent quant_state corruption."
+                )
```
Comment (SunMarc) [inline src/accelerate/utils/fsdp_utils.py]: then maybe we should warn the user to set the type for the storage to get the right memory saving

Comment (roycho96) [reply src/accelerate/utils/fsdp_utils.py]: Good point, updated the message to guide users toward the floating quant_storage path.

File: src/accelerate/utils/fsdp_utils.py [hunk 2]
```diff
@@ -657,14 +657,33 @@ def fsdp2_prepare_model(accelerator, model: torch.nn.Module) -> torch.nn.Module:
...
     for name, param in model.named_parameters():
         # this is a temporary fix whereby loading models with bnb params cannot be moved from
         # GPU to a meta device due with FSDP2 because torch operations don't return the original class type
         # bypassing the move to meta will still cause the VRAM spike, but at least it still will load
         if param.__class__.__name__ == "Params4bit":
-            model_has_params4bit = True
-            break
+            params4bit.append(param)
+
+    model_has_params4bit = len(params4bit) > 0
```
Comment (SunMarc) [inline src/accelerate/utils/fsdp_utils.py]: let's not append all param in params4bit. let's keep model_has_params4bit as we did before

File: tests/fsdp/test_fsdp.py [hunk 3]
```diff
@@ -471,6 +471,93 @@ def test_param_mapping_error_handling(self):
 
         AcceleratorState._reset_state(True)
 
+    def test_fsdp2_ignored_params_non_floating_params4bit(self):
```
Comment (SunMarc) [inline tests/fsdp/test_fsdp.py]: let's remove those tests, it's fine to not add for this as the tests are a bit too much over-engineered

Comment (roycho96) [reply tests/fsdp/test_fsdp.py]: removed

File: src/accelerate/utils/fsdp_utils.py (hunk 2 above)
Comment (roycho96) [reply src/accelerate/utils/fsdp_utils.py]: > let's not append all param in params4bit. let's keep model_has_params4bit as we did before

Done. Reverted to model_has_params4bit flag with early detection, and moved the incompatible filtering into the same loop to avoid a second pass over parameters.

Comment (SunMarc) [reply src/accelerate/utils/fsdp_utils.py]: thanks !
